### ISSUE TITLE
Scan speed in 0.12.4 terribly slow

### ISSUE THREAD

**Issue by vosscodes** [agersant/polaris#96]
Since upgrading, scans have become unbearably slow to the point that the app is unusable on my system when not much else is happening.

Debian sid on linux kernel 5.7.6 running polaris 0.12.4 in docker, scans have 4 threads capping out at about 20MB/s. Watching the logs shows no obvious errors or anything, but I can see very obvious hitching in the scan where all logging stops for a solid 20-30 seconds. The entire library is frozen up during a scan, and at these speeds, scans are taking longer than 30 minutes to finish.

Rolled back to 0.12.3, one thread is floating between 140-180MB/s with no issues using the app during the scan. I intended to look into this to see what I can do, but I really haven't had the time to put into this.

I'll try to investigate a bit in the next few days and report back.

**Comment by lnicola** [comment]
CC https://github.com/agersant/polaris/pull/86.

As a workaround, you can set the RAYON_NUM_THREADS (spelling?). environment variable to 1, but I'm not sure what's up. Even mechanical drives work best around QD 8 or so. Is there anything strange about your hardware or software?

**Comment by agersant** [comment]
I gathered more data on my own machine, not seeing the performance regression.
Re-index times on a warm cache of ~3200 songs, using a Windows machine with high-end CPU and SSD:

- Before #86 : 1,232,722,510 ns/iter (+/- 251,241,700)
- After #86 : 711,286,870 ns/iter (+/- 336,083,405)
- After #86, thread pool limited to one thread: 1,394,693,970 ns/iter (+/- 252,379,846)

**Comment by agersant** [comment]
The only odd thing I see in #86 that I missed in my original code review is [this line](https://github.com/agersant/polaris/pull/86/files#diff-5fa3fb1d7b39e2c72e9e3738ed482c5aR363) using `iter().par_bridge()` instead of `par_iter()` and I'm not sure what implications this can have (I imagine not much).

**Comment by vosscodes** [comment]
On 0.12.4 with RAYON_NUM_THREADS set to 1, the scan speed did improve somewhat and the app would load a bit better during a scan.
Leaving 0.12.4 up and letting it scan a few times with a warm cache definitely has better performance. I had been switching back and forth between the versions and probably seeing high times when nothing was cached.

That said, I'm still seeing a significant drop, and no music plays until it finishes. Loading up the library/random pages or searching simply don't return anything.

Watching a scan on 0.12.3 and I do see the logs stopping while reading files, but nothing longer than a few seconds here and there reading larger flac files.
Load average is a tiny bit higher at 1.91, memory usage is the same, and there are no issues playing music/reading the library.
```
polaris      | 19:04:53 [INFO] [src/index/update.rs:28] Library index update took 398.872 seconds
```

Scanning on 0.12.4 with RAYON_NUM_THREADS unset and no other I/O going on:
```
polaris      | 19:15:07 [INFO] [src/index/update.rs:28] Library index update took 791.015 seconds
```

The scan taking a bit longer to go through 13.5k files is honestly fine, but losing playback for almost 15 minutes is rough if you're scanning frequently.
strace shows it still doing its thing when logs stop, so that might be a red herring, but I'm seeing several minute gaps at times: https://gist.github.com/vosscodes/99a53881fcb5f751e98743c045d4e78b

Not a high end CPU by any means, Intel(R) Core(TM) i5-4460 CPU @ 3.20GHz, but even with video streaming and some other things going on in the background, load average is around 1.51 scanning on 0.12.4, and no issues show up in any other application. Memory has about 20GB free during all these scans. Assuming it's on my end, I was half expecting to see issues with the video or other read operations during a scan. Spinning disks, not SSDs, but no SMART warnings in logs or in skdump, drive status is good. Nothing special going on in the software, apart from being [dockerized](https://github.com/ogarcia/docker-polaris) I guess.

Not sure what to make of this one 🤔

**Comment by vosscodes** [comment]
One night this week (or next weekend), I'll try playing with #86 to see if I can nail down exactly what's causing a drop for me.

398 might have been on the high end, even. Dropping back to 0.12.3 for now and the first scan was faster:
```
polaris      | 20:14:07 [INFO] [src/index/update.rs:27] Library index update took 337.637 seconds
```

**Comment by agersant** [comment]
Good news, I can also reproduce this with the Polaris install on my Raspberry Pi 4.

On `master` (Rocket), all requests that need to touch the filesystem stall during the index process. On the `actix` branch, every request everything does.

**Comment by lnicola** [comment]
@agersant can you confirm that not using `rayon` is faster (because that's pretty much what this issue is about)?

**Comment by agersant** [comment]
I wasn't precise enough in my previous message. I mean I can reproduce the issue of unusable app, I have not looked at indexing speeds. I'll get some with/without rayon comparison of stalls and indexing speed later this week.

**Comment by agersant** [comment]
Will edit this post as I go:

rayon indexing|cold boot index duration|usable during index
-|-|-
yes||no
no||

**Comment by lnicola** [comment]
@agersant you might want to run `echo 3 | sudo tee /proc/sys/vm/drop_caches` between tests (unless "cold boot" is really a reboot).

**Comment by agersant** [comment]
I'm using Balena OS reboots, which I think are close enough to an actual unplugging of the power cable:

https://www.balena.io/docs/learn/manage/actions/#reboot

**Comment by agersant** [comment]
Results are in in the table above. It matches the original observations by @vosscodes.

I'm not familiar with Linux perf tooling to profile what is actually happening under the hood. My guess is that in the rayon version:
- Client is unusable because the indexing is making thousands IO requests at a high-rate, from many threads. IO calls for indexing outnumber calls for request handlers by a huge factor, and the scheduler ends up mostly scheduling work for the indexing job as a result. This is somewhat confirmed by the fact that it is sometimes possible to browse the collection a bit (small amount of IO required for completion) but thumbnails never show up (lots of IO required for completion).
- Indexing is slower because of context switch overhead between the rayon threads, or because doing IO on multiple files at once is harming the disk cache mechanisms, or a little bit of both.

**Comment by lnicola** [comment]
Can you do a quick run with `RAYON_NUM_THREADS=1`

**Comment by agersant** [comment]
The [HDD](https://www.amazon.com/gp/product/B00671E8DG/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) supports USB3 and is plugged in to a USB3 port of the Pi. I have not verified what it is actually using.

Going to run one test with `RAYON_NUM_THREADS=1` (and then go to bed).

**Comment by agersant** [comment]
Table above updated.

**Comment by lnicola** [comment]
Thanks. So it looks like the indexing speed regression can be mitigated by setting that env variable, but the API is still starved.

Final question (for tomorrow): what file system are you using on the HDD?

**Comment by lnicola** [comment]
So I thought I'd try again. It's still a Celeron J1900 with a 2-HDD ZFS mirror (SATA 2 + SATA 3). One difference is that I have a bit of L2ARC (SSD cache) which is now persistent.

```
# rayon, drop caches
Library index update took 40.458 seconds
Library index update took 39.421 seconds
Library index update took 40.492 seconds

# sequential, drop caches
Library index update took 77.108 seconds
Library index update took 75.11 seconds
Library index update took 72.321 seconds

# rayon, reboot
Library index update took 38.785 seconds
Library index update took 38.468 seconds
Library index update took 37.633 seconds

# sequential, reboot
Library index update took 94.948 seconds
Library index update took 264.439 seconds
Library index update took 125.87 seconds
```

**Comment by vosscodes** [comment]
sorry to say I didn't have much time to look at this further, but observed similar results to @agersant in his table above when testing. when `RAYON_NUM_THREADS=1`, audio and navigation are sometimes usable but thumbnails never are. scan speed is improved, but not by a large amount.

thanks for following up on this. I should have more time in the coming weeks to help poke at things.
btw, the [disk in question](https://www.amazon.com/Toshiba-Performance-Desktop-Internal-HDWE150XZSTA/dp/B013JPLKQK) is ext4 and using a sata connection.

**Comment by agersant** [comment]
> Final question (for tomorrow): what file system are you using on the HDD?

ext4

**Comment by lnicola** [comment]
Oof. I guess we can make this a run-time configuration.

**Comment by agersant** [comment]
I'm not too thrilled about that perspective. The option would be hard for users to understand and evaluate, with a possibility of the client becoming unusable when they try to test it. Also of note, I found that the `rayon` docs explicitly recommend not using it to parallelize IO-heavy work (oops).

I spent tonight re-writing the indexer with a fairly different approach to the existing one. I split the work in three threads, one doing exclusively disk traversal and metadata reading, one computing what needs to go in the database, and one making inserts in the DB. In my brief testing so far, I am getting numbers on-par with the better ones in the table above. The client is fully usable during indexing too.

@lnicola Could you do some testing on your setup using the `rocket-no-rayon-index` (1b407f9987bb04e6aa3b17b10eebfef7af3a18b6)? I am curious to see where the index timing lands for you.

(@vosscodes also welcome to do some testing but I assume our configurations are fairly similar, considering results so far)

**Comment by lnicola** [comment]
It's still slower for me, but not as slow as the sequential version:

```
# new indexer, drop caches
Library index update took 77.053 seconds
Library index update took 64.889 seconds
Library index update took 64.199 seconds

# new indexer, reboot
Library index update took 80.036 seconds
Library index update took 64.589 seconds
Library index update took 66.054 seconds
```

**Comment by agersant** [comment]
This looks like excellent news! I am hoping to get it a little lower but I would already call this shippable.

I am trying to figure a good way to profile which of the three threads is the bottleneck (stay tuned). I suspect in your case it's the `collector` thread. If that is correct, it shouldn't be too hard to parallelize its work further.

> The FLAC metadata contains the duration, but IIRC MP3s don't so they need to be read. 
It depends on the MP3. They can have an optional [XING header](https://www.codeproject.com/Articles/8295/MPEG-Audio-Frame-Header#XINGHeader) which has enough information to compute the duration even for VBR files. Most of the mp3 I am testing with should have this set up correctly.

**Comment by agersant** [comment]
A brief investigation into CPU profiling looks like `cargo flamegraph` may be the way to go. Unfortunately it seems broken on Windows (even though it is supported). Not sure how valuable the output will be considering this also doesn't plot work on a time axis.

Maybe I'll just write the changes parallelizing the `collector` work so you can simply run it and compare.

**Comment by lnicola** [comment]
Here is my `zfs iostat` output over intervals of one second for two runs:

```
# new indexer, 68 s
              capacity     operations     bandwidth 
pool        alloc   free   read  write   read  write
----------  -----  -----  -----  -----  -----  -----
tank         820G  1.01T      0      0      0      0
tank         820G  1.01T      0      0      0      0
tank         820G  1.01T     62      0  4.38M      0
tank         820G  1.01T    123      0  27.6M      0
tank         820G  1.01T    296      0  77.0M      0
tank         820G  1.01T     37      0  6.87M      0
tank         820G  1.01T     13      0  1.28M      0
tank         820G  1.01T     51      0  4.86M      0
tank         820G  1.01T     49      0  4.42M      0
tank         820G  1.01T     52      0  5.15M      0
tank         820G  1.01T     66      0  8.21M      0
tank         820G  1.01T    310      0  59.6M      0
tank         820G  1.01T    219      0  47.1M      0
tank         820G  1.01T    138      0  31.8M      0
tank         820G  1.01T     24      0  4.98M      0
tank         820G  1.01T    227      0  47.7M      0
tank         820G  1.01T     36      0  5.36M      0
tank         820G  1.01T    101      0  71.7M      0
tank         820G  1.01T     78      0  10.1M      0
tank         820G  1.01T      0      0      0      0
tank         820G  1.01T     25      0  4.11M      0
tank         820G  1.01T    315      0  62.0M      0
tank         820G  1.01T    318      0  55.9M      0
tank         820G  1.01T    354      0  74.0M      0
tank         820G  1.01T    219      0  46.5M      0
tank         820G  1.01T    271      0  85.5M      0
tank         820G  1.01T    333      0  94.5M      0
tank         820G  1.01T    250      0  65.2M      0
tank         820G  1.01T      9      0  4.73M      0
tank         820G  1.01T    133      0  25.9M      0
tank         820G  1.01T     83      0  16.3M      0
tank         820G  1.01T      0      0      0      0
tank         820G  1.01T    144      0  22.1M      0
tank         820G  1.01T    286      0  44.8M      0
tank         820G  1.01T    316      0  61.9M      0
tank         820G  1.01T    248      0  53.7M      0
tank         820G  1.01T    295      0  98.1M      0
tank         820G  1.01T    307      0  77.7M      0
tank         820G  1.01T    333      0  71.3M      0
tank         820G  1.01T    396      0  73.6M      0
tank         820G  1.01T    401      0  73.3M      0
tank         820G  1.01T    322      0  72.8M      0
tank         820G  1.01T    298      0  87.4M      0
tank         820G  1.01T    222      0  85.3M      0
tank         820G  1.01T    240      0  67.2M      0
tank         820G  1.01T     85      0  20.7M      0
tank         820G  1.01T    188      0  52.9M      0
tank         820G  1.01T    254      0  79.7M      0
tank         820G  1.01T    267      0  83.6M      0
tank         820G  1.01T    247      0  82.5M      0
tank         820G  1.01T    257      0  65.3M      0
tank         820G  1.01T    238      0  64.6M      0
tank         820G  1.01T    254      0  57.7M      0
tank         820G  1.01T    279      0  74.3M      0
tank         820G  1.01T    253      0  77.8M      0
tank         820G  1.01T    248      0  80.3M      0
tank         820G  1.01T    253      0  64.8M      0
tank         820G  1.01T    253      0  69.6M      0
tank         820G  1.01T    193      0  38.6M      0
tank         820G  1.01T    110      0  14.9M      0
tank         820G  1.01T    170      0  36.0M      0
tank         820G  1.01T    128      0  21.2M      0
tank         820G  1.01T     93      0  19.8M      0
tank         820G  1.01T    108      0  22.7M      0
tank         820G  1.01T    138      0  25.6M      0
tank         820G  1.01T    127      0  30.7M      0
tank         820G  1.01T    126      0  27.0M      0
tank         820G  1.01T    109      0  23.9M      0
tank         820G  1.01T     99      0  19.1M      0
tank         820G  1.01T     11      0  1.99M      0
tank         820G  1.01T      0      0      0      0
tank         820G  1.01T      0      0      0      0

# master, 38 s
              capacity     operations     bandwidth 
pool        alloc   free   read  write   read  write
----------  -----  -----  -----  -----  -----  -----
tank         820G  1.01T      0      0   127K      0
tank         820G  1.01T      0      0      0      0
tank         820G  1.01T     75      0  11.2M      0
tank         820G  1.01T    319      0  83.8M      0
tank         820G  1.01T    284      0  95.4M      0
tank         820G  1.01T    255      0   112M      0
tank         820G  1.01T    252      0   101M      0
tank         820G  1.01T    278      0  93.7M      0
tank         820G  1.01T    214      0  90.4M      0
tank         820G  1.01T    222      0  74.8M      0
tank         820G  1.01T    278      0  70.0M      0
tank         820G  1.01T    232      0  49.2M      0
tank         820G  1.01T    283      0  76.9M      0
tank         820G  1.01T    251      0  92.3M      0
tank         820G  1.01T    241      0  81.6M      0
tank         820G  1.01T    183      0  59.9M      0
tank         820G  1.01T    234      0  79.8M      0
tank         820G  1.01T    227      0  77.8M      0
tank         820G  1.01T    239      0  90.4M      0
tank         820G  1.01T    197      0  79.0M      0
tank         820G  1.01T    257      0  87.9M      0
tank         820G  1.01T    260      0  93.6M      0
tank         820G  1.01T    265      0  83.1M      0
tank         820G  1.01T    255      0  79.8M      0
tank         820G  1.01T    275      0  74.0M      0
tank         820G  1.01T    223      0  67.8M      0
tank         820G  1.01T    240      0  71.1M      0
tank         820G  1.01T    256      0  71.7M      0
tank         820G  1.01T    237      0  71.6M      0
tank         820G  1.01T    243      0  79.7M      0
tank         820G  1.01T    210      0   111M      0
tank         820G  1.01T    230      0  72.3M      0
tank         820G  1.01T    226      0  74.0M      0
tank         820G  1.01T    239      0  83.5M      0
tank         820G  1.01T    224      0  69.9M      0
tank         820G  1.01T    190      0  68.7M      0
tank         820G  1.01T    196      0  68.8M      0
tank         820G  1.01T    216      0  87.0M      0
tank         820G  1.01T    223      0  69.9M      0
tank         820G  1.01T    169      0  49.9M      0
tank         820G  1.01T      0      0      0      0
```

So it looks like the new implementation sometimes can't feed enough read requests to the disk.

**Comment by lnicola** [comment]
Here is some `perf` output for that thread (or process, I'm not really sure):

```
   6.78%  [kernel]                   [k] memcpy_erms
   5.16%  libc-2.32.so               [.] __memmove_sse2_unaligned_erms
   4.46%  [kernel]                   [k] clear_page_erms
   3.07%  [kernel]                   [k] copy_user_enhanced_fast_string
   2.75%  polaris                    [.] mp3_duration::from_file
   2.66%  polaris                    [.] <std::io::buffered::bufreader::BufReader<R> as std::io::Read>::read
   2.64%  [zfs]                      [k] lz4_decompress_zfs
   1.78%  [kernel]                   [k] preempt_count_add
   1.51%  [kernel]                   [k] mutex_lock
   1.49%  libc-2.32.so               [.] _int_malloc
   1.47%  polaris                    [.] std::io::Read::read_exact
   1.09%  libc-2.32.so               [.] _int_free
   1.00%  [kernel]                   [k] memset_erms
   0.99%  polaris                    [.] std::io::copy::generic_copy
   0.98%  polaris                    [.] mp3_duration::context::Context<T>::read_exact
   0.90%  libc-2.32.so               [.] malloc
   0.88%  [kernel]                   [k] _raw_spin_lock
   0.88%  [kernel]                   [k] mutex_unlock
   0.85%  [zfs]                      [k] zio_create
   0.79%  [zfs]                      [k] dbuf_rele_and_unlock
   0.77%  [kernel]                   [k] kmem_cache_alloc
   0.73%  [kernel]                   [k] __kmalloc_node
   0.67%  [kernel]                   [k] preempt_count_sub
   0.64%  [kernel]                   [k] kmem_cache_free
   0.63%  [kernel]                   [k] __fsnotify_parent
   0.61%  [zfs]                      [k] dmu_buf_hold_array_by_dnode
   0.61%  [kernel]                   [k] kfree
   0.59%  polaris                    [.] metaflac::block::VorbisComment::from_bytes
   0.56%  [zfs]                      [k] dbuf_hold_impl
   0.52%  [kernel]                   [k] in_lock_functions
   0.52%  [spl]                      [k] spl_kmem_cache_alloc
   0.51%  [zfs]                      [k] zio_wait
   0.48%  [zfs]                      [k] zio_done
   0.47%  libc-2.32.so               [.] realloc
   0.46%  libc-2.32.so               [.] cfree@GLIBC_2.2.5
```

**Comment by agersant** [comment]
Thanks for the data, these are very interesting. My guess was wrong and looks like `traverser` is the bottleneck then, not `collector`.

What surprises me the most is that the bandwidth peaks around the same high of 95MBps but is much lower on average with the new indexer. There's even a few seconds with no data being read at all?? That thread does basically nothing except reading the disk so I don't understand what could be keeping it busy for an entire second at any point.

**Comment by lnicola** [comment]
That's the IO that actually hits the storage. There might be some prefetching or caching involved.

**Comment by lnicola** [comment]
The [`ignore`](https://docs.rs/ignore/) crate has a parallel file system iterator that might be worth trying.

**Comment by agersant** [comment]
I tried using the `ignore` crate but its API does not fir our needs very well. It uses a visitor pattern for individual files, but Polaris needs to have eyes on all files within a directory at the same time (to determine metadata for the directory itself).

Instead, I multi-threaded the `traverser` work myself, by closely referencing the `ignore` implementation. The results have broken all speed records on the Raspeberry Pi setup and one my dev machine. I updated the table in the comment above with these new results. Client is still very responsive throughout so I except this is the version that will ship.

This variant defaults to using 4 threads for the `traverser`, unless the machine has less than 4 logical cores. This number can be manually set with the `POLARIS_NUM_TRAVERSER_THREADS`. I picked 4 as the default because that's the number after which I stopped getting significantly better times when testing on the Windows machine.

**Comment by lnicola** [comment]
```
# new indexer (take 2), drop caches
Library index update took 92.975 seconds
Library index update took 85.283 seconds
Library index update took 79.178 seconds

# new indexer (take 2), reboot
Library index update took 104.042 seconds
Library index update took 55.745 seconds
Library index update took 96.143 seconds
```

So it got even slower for me.

**Comment by agersant** [comment]
Well, it did fix the issue of the disk not being used to read a consistently high amount of data (assuming the column measured in Megabytes in `zpool iostat` is the bandwidth column in your prior `zfs stat`).

Can you try some variations on the number of threads using `POLARIS_NUM_TRAVERSER_THREADS`? 

I'm even more lost than before. It's especially weird how much the timings fluctuate in this last test. From 55s to 104s with the same code is very high variance. For the `perf` captures, can you use the `-g` option (possibly with  --max-stack 6) so it's a little clearer what the code is busy doing?

I'm not a fan of the mutex and sleep but that's what used in ripgrep/ignore so I figured it was good enough for me too. I originally used a crossbeam channel for this but I drove myself into a corner because I didn't have the `num_pending` atomic. I'll try again now that I have this extra wisdom - although I don't expect a big perf difference.

**Comment by agersant** [comment]
I'm having issues deploying it to my Raspberry Pi, but if you want to try the version without the mutex/sleep channel, it's in eb39835e2a9201ba429f3c85437c2776fb829895.

**Comment by lnicola** [comment]
I guess I should just concede that my system is behaving in an unusual way, perhaps due to the particularities of ZFS.

```
# crossbeam_channel, drop caches
Library index update took 108.789 seconds
Library index update took 93.615 seconds
Library index update took 89.814 seconds

# crossbeam_channel, reboot
Library index update took 94.242 seconds
Library index update took 71.395 seconds
Library index update took 73.828 seconds
```

**Comment by agersant** [comment]
Well, at the very least we got a new indexer out of this - which looks better than the previous one and solves stalling issues.
Merged in https://github.com/agersant/polaris/pull/107

**Comment by agersant** [comment]
Fixed in release 0.13
